test/test_puma_server.rb

```diff
@@ -1647,6 +1647,41 @@ def test_force_shutdown
     shutdown_requests(**opts, post: true, s2_response: /408/)
   end
 
+  def test_shutdown_with_shutdown_debug
+    out, _ = capture_io do
+      shutdown_requests(s1_response: /204/, s2_response: /204/, shutdown_debug: true)
+    end
+
+    assert_equal 1, out.scan("Shutdown initiated").length
+    assert_equal 1, out.scan("Begin thread backtrace dump").length
+  end
+
+  def test_force_shutdown_with_shutdown_debug
+    out, _ = capture_io do
+      shutdown_requests(s1_complete: false, s1_response: /503/, s2_response: nil, shutdown_debug: true, force_shutdown_after: 0)
+    end
+
+    assert_equal 1, out.scan("Shutdown initiated").length
+    assert_equal 1, out.scan("Begin thread backtrace dump").length
+  end
+
+  def test_shutdown_with_shutdown_debug_on_force
+    out, _ = capture_io do
+      shutdown_requests(s1_response: /204/, s2_response: /204/, shutdown_debug: :on_force)
+    end
+
+    assert_equal 0, out.scan("Begin thread backtrace dump").length
+  end
+
+  def test_force_shutdown_with_shutdown_debug_on_force
+    out, _ = capture_io do
+      shutdown_requests(s1_complete: false, s1_response: /503/, s2_response: nil, shutdown_debug: :on_force, force_shutdown_after: 0)
+    end
+
+    assert_equal 1, out.scan("Shutdown timeout exceeded").length
+    assert_equal 1, out.scan("Begin thread backtrace dump").length
```

Comment:
I can't think of an easy way to assert the second back-trace when the shutdown grace timeout has exceeded as it's not configurable, but I think this is sufficient anyway.